fastlane/fastlane
# [spaceship][deliver] add app clips support for specifying default experience metadata

### Checklist


### Motivation and Context

There is currently no app clips functionality within fastlane. My team and I ran into the issue described in #17639 and had to do manual releases (not fun). I decided to tackle this to help lessen the barrier a bit for others utilizing app clips.

There existed another seemingly abandoned WIP PR #19043 that initially implemented app clips using the web session and private APIs. The ASC API then came out with full app clips support, but with changed models/api-endpoints. This in turn required that PR to re-implement much of it's content, which has not yet been done.

This PR looks to implement the goals of #19043, but with additional `deliver` support so developers can immediately start working with app clips metadata automation via simple `Fastfile` updates.

Resolves #17639

### Description

This PR is one of a few app clips related changes I have made. I am presenting the PRs in smaller pieces so the implications of each PR can be discussed in a tighter scope.

The following app clip default experience functionality has been added to `deliver`:
- option to specify action (PLAY, VIEW, ...)
- option to specify localized subtitle
- option to specify localized default experience header image

If none of these options are specified, then fastlane will copy over the app clip default experience metadata from the live version. This is similar to the default behavior of the ASC UI discussed [here](https://github.com/fastlane/fastlane/issues/17639#issuecomment-1130196397).

Example:
```rb
deliver(
  app_clip_default_experience_action: "PLAY",
  app_clip_default_experience_subtitle: {
    "en-US": "this is a test from fastlane",
    "es-MX": "está es una prueba de fastlane"
  },
  app_clip_header_images_path: "./fastlane/app-clip-header-images"
)
```

The localized header images are read from the filesystem same as screenshots:

```sh
$ tree app-clip-header-images/
app-clip-header-images
├── en-US
│   └── random-header-image.jpg
└── es-MX
    └── another-header-image.jpg

2 directories, 2 files
```

These options relate to the ASC UI in the "App Clips" section under a specific App Store version: ![CleanShot 2022-05-19 at 12 12 29@2x](https://user-images.githubusercontent.com/11725484/169361521-73063d31-82ac-44b0-aeab-ae5dcf9b5907.png)

### Discussion

Prior discussion can be [found here](https://github.com/fastlane/fastlane/issues/17639#issuecomment-1121590421).

This PR does NOT implement anything wrt app clip **advanced experiences**. This PR focuses on the default experience associated with the App Store versions only. The advanced experiences are a bit of different beast. If there's interest and it makes sense, I may consider adding that in another PR since I am now familiar with the ASC APIs.

The following should be addressed before this PR is ready to be merged:

1. ~~Apple allows more than one app clip target per app, this PR uses the [first listed app clip only](https://github.com/qatonda/fastlane/blob/9bf2f6bd8bf8dd2bef97dca0e11c6a44928c79a5/deliver/lib/deliver/upload_app_clip_default_experience_metadata.rb#L26).~~
  - Resolved in [this comment](https://github.com/fastlane/fastlane/pull/20300#issuecomment-1133162510)

<a href="#discussion-point-2" id="discussion-point-2">#</a> 2.  ~~docs: I see [generated documentation here](https://github.com/fastlane/docs/tree/master/docs/generated), but it's unclear how I update them. I would be happy to update the docs and add an app clips section if someone pointed me in the right direction.~~
  - Resolved with [a commit to update docs](https://github.com/fastlane/fastlane/pull/20300/commits/de5f6d90cc72a7506bc19120cee64a21d4a7c4d4)

3. Do we want the app clips metadata to show up in the interactive HTML metadata report?
    - I have not ever used this feature, but I noticed it while testing.

#### Code Quality

> Disclaimer: I am not a ruby programmer, so I will have missed a lot of ruby-isms in this PR, but I tried my best to keep the code style consistent with the rest of the codebase.

I have not yet added tests for these changes. I have tested them using a small test project and my own projects in production.

Some of the code in this PR has been copied from other relevant parts of the codebase.

##### For example:

The new file [`spaceship/connect_api/models/app_clip_header_image.rb`](https://github.com/fastlane/fastlane/blob/55fb13119ff9b65c865977be6390ecd6cd738360/spaceship/lib/spaceship/connect_api/models/app_clip_header_image.rb) has code copied from [`.../models/app_screenshot.rb`](https://github.com/fastlane/fastlane/blob/master/spaceship/lib/spaceship/connect_api/models/app_screenshot.rb) since the upload functionality was the same. I decided not to inherit or subclass because I was sure that making the upload process more general would be more work than it was worth. I am absolutely open to hearing counter arguments.

The PR #17639 has a [similar comment about extracting common functionality here](https://github.com/fastlane/fastlane/pull/19043#pullrequestreview-698633451).

### Testing Steps

#### Basic Testing Setup

1. Update `Gemfile` to 👇 and run `bundle install`, `bundle update fastlane`, or `bundle update`
```rb
gem "fastlane", :git => 'https://github.com/qatonda/fastlane.git', :ref => "matt/app-clips-default-experience"
```
2. Upload a binary to ASC with an app clip target
    - you can follow [a project tutorial provided by Apple here](https://developer.apple.com/documentation/swiftui/fruta_building_a_feature-rich_app_with_swiftui)

#### What to Test

Before specifying any `deliver` options:
- Confirm that fastlane uses the live version's app clip metadata for the current edit version
    - It should copy over the action, subtitle, and header image of the live version
    - This is not already the default when the AS version is created via the ASC APIs, thus the following issue #17639

After specifying the app clip default experience `deliver` options:
- Confirm the app clip default experience metadata specified in the `deliver` options reflects the ASC UI after running `deliver`
- Confirm that fastlane enforces that the app clip header image is exactly 1800x1200 in size
- Confirm that fastlane enforces only one app clip header image per locale
- Confirm that the additional fastlane messages via `UI.{message,important,...}` are appropriate
